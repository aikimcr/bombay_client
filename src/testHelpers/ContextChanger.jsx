import React, { useState } from 'react';

import { BombayLoginContext } from '../Context/BombayLoginContext';
import { MemoryRouter } from 'react-router';

// Use this to turn the crank on context changes.
export const ContextChanger = ({ initialLoggedIn, children }) => {
  const [loggedIn, setLoggedIn] = useState(initialLoggedIn);

  const loginContext = {
    loggedIn,
    setLoggedIn,
  };

  function toggleLogin() {
    loginContext.setLoggedIn((oldLoggedIn) => !oldLoggedIn);
  }

  return (
    <BombayLoginContext.Provider value={loginContext}>
      <MemoryRouter>
        {children}
        <button
          className="change-test-login"
          data-testid="change-test-login"
          onClick={toggleLogin}
        >
          Change Login
        </button>
      </MemoryRouter>
    </BombayLoginContext.Provider>
  );
};
