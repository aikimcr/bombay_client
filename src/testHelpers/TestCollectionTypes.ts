import {
  CollectionBase,
  CollectionBaseConstructorArgs,
} from '../Model/CollectionBase';
import {
  ModelBase,
  ModelBaseConstructorArgs,
  ModelDataBase,
} from '../Model/ModelBase';

export const TestCollectionOneURL = 'http://localhost:2001/testone';

export interface TestModelOneData extends ModelDataBase {
  description: string;
}

export class TestModelOne extends ModelBase<TestModelOneData> {
  static TableName = 'testone';

  constructor(args: ModelBaseConstructorArgs<TestModelOneData>) {
    super({
      ...args,
      tableName: TestModelOne.TableName,
    });
  }

  get description() {
    return this._data.description;
  }

  set description(newDescription: string) {
    this._data.description = newDescription;
  }

  protected isModelData(data: object): data is TestModelOneData {
    return super.isModelData(data) && Object.hasOwn(data, 'description');
  }
}

export class TestCollectionOne extends CollectionBase<
  TestModelOneData,
  TestModelOne
> {
  constructor(
    args: CollectionBaseConstructorArgs<TestModelOneData, TestModelOne>,
  ) {
    super({
      ...args,
      tableName: TestModelOne.TableName,
    });
  }

  protected createAModelFromDef(def: TestModelOneData): TestModelOne {
    return TestModelOne.from<TestModelOneData, TestModelOne>(def, {
      keepId: true,
    });
  }
}

export interface TestCollectionOneFetchBody {
  data: TestModelOneData[];
  nextPage?: string;
  prevPage?: string;
}

export interface TestCollectionOneFetchResult {
  body: TestCollectionOneFetchBody;
}
