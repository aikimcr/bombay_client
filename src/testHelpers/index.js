export * from './ContextChanger';
export * from './modelTools';
