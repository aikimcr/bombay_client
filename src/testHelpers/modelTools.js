// Import this after all mocks.  Especially Network and login mocks.

import casual from "casual";

import * as Network from "../Network/Network";

import ModelBase from "../Model/ModelBase";
import ArtistModel from "../Model/ArtistModel";
import SongModel from "../Model/SongModel";

const allNames = {};
casual.define("uniqueName", function (nameType, clear) {
  if (!allNames[nameType]) allNames[nameType] = new Set();

  const oldSize = allNames[nameType].size;
  let newName;

  while (allNames[nameType].size === oldSize) {
    newName = casual.full_name;
    allNames[nameType].add(newName);
  }

  return newName;
});

const idSequences = {};
casual.define("nextId", function (sequenceName) {
  const nextOne = (idSequences[sequenceName] ?? 0) + 1;
  idSequences[sequenceName] = nextOne;
  return nextOne;
});

export function makeAModel(tableName = "table1", fieldsCallback) {
  const def = {};
  def.id = casual.nextId(tableName);
  def.name = casual.uniqueName(tableName);

  if (fieldsCallback) {
    fieldsCallback(def);
  }

  let modelClass = ModelBase;

  switch (tableName) {
    case "artist":
      modelClass = ArtistModel;
      break;

    case "song":
      modelClass = SongModel;
      def.key_signature = "C";
      def.tempo = 120;
      def.lyrics = "O Solo Mio! The troubles I have seen";

      if (def.artist) {
        def.artist_id = def.artist.id;
      } else {
        const [artist] = makeAModel("artist");
        def.artist_id = artist.id;
        def.artist = artist;
      }
      break;

    default:
  }

  def.url = Network.buildURL({ path: `/${tableName}/${def.id}` });

  return [def, modelClass.from(def)];
}

export function makeModels(
  length = 10,
  query = {},
  tableName = "table1",
  fieldsCallback,
) {
  const result = {
    data: [],
  };

  const models = [];

  while (result.data.length < length) {
    const [def, model] = makeAModel(tableName, fieldsCallback);
    result.data.push(def);
    models.push(model);
  }

  let offset = (query.offset || 0) - length;
  let limit = query.limit || length;

  if (offset >= 0) {
    result.prevPage = Network.prepareURLFromArgs(tableName, {
      offset,
      limit,
    }).toString();
  }

  if (limit <= length) {
    offset = (query.offset || 0) + length;
    result.nextPage = Network.prepareURLFromArgs(tableName, {
      offset,
      limit,
    }).toString();
  }

  return [result, models];
}
