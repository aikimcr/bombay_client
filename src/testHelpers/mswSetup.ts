import { http, HttpResponse } from 'msw';
import {
  TestCollectionOneFetchBody,
  TestCollectionOneURL,
} from './TestCollectionTypes';

const TestData: TestCollectionOneFetchBody[] = [
  {
    data: [
      {
        id: 1,
        name: 'special voice melted',
        description: 'Ipsum amet consectetur commodo et ut et voluptate sint.',
      },
      {
        id: 2,
        name: 'picture than nails',
        description:
          'Reprehenderit aute ullamco quis qui duis do et quis minim labore esse.',
      },
      {
        id: 3,
        name: 'never ask dead',
        description:
          'Minim dolor eu fugiat tempor irure labore nostrud pariatur commodo aute.',
      },
      {
        id: 4,
        name: 'seems steady cry',
        description: 'Mollit voluptate nostrud nulla laborum nisi do occaecat.',
      },
      {
        id: 5,
        name: 'jack southern jet',
        description: 'Eiusmod qui culpa veniam eiusmod adipisicing aliqua.',
      },
      {
        id: 6,
        name: 'manner by trap',
        description:
          'Pariatur sit nulla laboris minim proident dolor aute amet id magna.',
      },
      {
        id: 7,
        name: 'perfect canal arrow',
        description:
          'Labore laboris magna reprehenderit aliquip ex Lorem sit non nisi exercitation ipsum voluptate.',
      },
      {
        id: 8,
        name: 'may river begun',
        description: 'Officia ipsum occaecat ad proident elit aliqua.',
      },
      {
        id: 9,
        name: 'donkey national writing',
        description:
          'Pariatur deserunt duis dolore velit cupidatat sunt ex duis duis veniam aliqua consectetur laborum.',
      },
      {
        id: 10,
        name: 'frog private would',
        description:
          'Commodo reprehenderit aute adipisicing esse nostrud pariatur incididunt eu.',
      },
    ],
    nextPage: `${TestCollectionOneURL}?offset=10&limit=10`,
  },
  {
    data: [
      {
        id: 11,
        name: 'notice visit happily',
        description: 'Ipsum amet consectetur commodo et ut et voluptate sint.',
      },
      {
        id: 12,
        name: 'condition aid hung',
        description:
          'Reprehenderit aute ullamco quis qui duis do et quis minim labore esse.',
      },
      {
        id: 13,
        name: 'variety class later',
        description:
          'Minim dolor eu fugiat tempor irure labore nostrud pariatur commodo aute.',
      },
      {
        id: 14,
        name: 'been edge student',
        description: 'Mollit voluptate nostrud nulla laborum nisi do occaecat.',
      },
      {
        id: 15,
        name: 'his brief length',
        description: 'Eiusmod qui culpa veniam eiusmod adipisicing aliqua.',
      },
      {
        id: 16,
        name: 'flame value fort',
        description:
          'Pariatur sit nulla laboris minim proident dolor aute amet id magna.',
      },
      {
        id: 17,
        name: 'police electric found',
        description:
          'Labore laboris magna reprehenderit aliquip ex Lorem sit non nisi exercitation ipsum voluptate.',
      },
      {
        id: 18,
        name: 'letter beneath day',
        description: 'Officia ipsum occaecat ad proident elit aliqua.',
      },
      {
        id: 19,
        name: 'mix leaving worse',
        description:
          'Pariatur deserunt duis dolore velit cupidatat sunt ex duis duis veniam aliqua consectetur laborum.',
      },
      {
        id: 20,
        name: 'younger simplest swept',
        description:
          'Commodo reprehenderit aute adipisicing esse nostrud pariatur incididunt eu.',
      },
    ],
    nextPage: `${TestCollectionOneURL}?offset=20&limit=10`,
    prevPage: `${TestCollectionOneURL}?offset=0&limit=10`,
  },
  {
    data: [
      {
        id: 21,
        name: 'barn mixture handle',
        description: 'Ipsum amet consectetur commodo et ut et voluptate sint.',
      },
      {
        id: 22,
        name: 'pair arm temperature',
        description:
          'Reprehenderit aute ullamco quis qui duis do et quis minim labore esse.',
      },
      {
        id: 23,
        name: 'whom answer try',
        description:
          'Minim dolor eu fugiat tempor irure labore nostrud pariatur commodo aute.',
      },
      {
        id: 24,
        name: 'there father bread',
        description: 'Mollit voluptate nostrud nulla laborum nisi do occaecat.',
      },
      {
        id: 25,
        name: 'laid correct lack',
        description: 'Eiusmod qui culpa veniam eiusmod adipisicing aliqua.',
      },
      {
        id: 26,
        name: 'factor willing dead',
        description:
          'Pariatur sit nulla laboris minim proident dolor aute amet id magna.',
      },
      {
        id: 27,
        name: 'date chair free',
        description:
          'Labore laboris magna reprehenderit aliquip ex Lorem sit non nisi exercitation ipsum voluptate.',
      },
      {
        id: 28,
        name: 'love mad nearest',
        description: 'Officia ipsum occaecat ad proident elit aliqua.',
      },
      {
        id: 29,
        name: 'truth eleven differ',
        description:
          'Pariatur deserunt duis dolore velit cupidatat sunt ex duis duis veniam aliqua consectetur laborum.',
      },
      {
        id: 30,
        name: 'younger describe eight',
        description:
          'Commodo reprehenderit aute adipisicing esse nostrud pariatur incididunt eu.',
      },
    ],
    prevPage: `${TestCollectionOneURL}?offset=10&limit=10`,
  },
];

export const TestCollectionHandlers = [
  http.get(TestCollectionOneURL, ({ request }) => {
    const requestUrl = new URL(request.url);

    if (requestUrl.searchParams.size > 0) {
      const offset = requestUrl.searchParams.get('offset');
      if (offset === '10') {
        return HttpResponse.json(TestData[1]);
      }

      if (offset === '20') {
        return HttpResponse.json(TestData[2]);
      }
    }

    return HttpResponse.json(TestData[0]);
  }),
];
