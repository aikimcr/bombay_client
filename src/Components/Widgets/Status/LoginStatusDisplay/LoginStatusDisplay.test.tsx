import React from 'react';
import { act, render, screen } from '@testing-library/react';

import {
  mockLoginStatus,
  mockRefreshToken,
  mockLogin,
  mockLogout,
} from '../../../../Network/testing';

jest.mock('../../../../Network/Login', () => {
  const originalModule = jest.requireActual('../../../../Network/Login');

  return {
    __esModule: true,
    ...originalModule,
    loginStatus: mockLoginStatus,
    refreshToken: mockRefreshToken,
    login: mockLogin,
    logout: mockLogout,
  };
});

import {
  mockUseRouteManager,
  mockNavigateToRoute,
} from '../../../../Hooks/testing/mocks/mockUseRouteManager';

jest.mock('../../../../Hooks/useRouteManager', () => {
  const originalModule = jest.requireActual(
    '../../../../Hooks/useRouteManager',
  );

  return {
    __esModules: true,
    ...originalModule,
    useRouteManager: mockUseRouteManager,
  };
});

import { ContextChanger } from '../../../../testHelpers/ContextChanger';

import { LoginStatusDisplay } from '.';

describe('LoginStatusDisplay', () => {
  describe('when logged out', () => {
    it('should show the login button', async () => {
      const { asFragment } = render(
        <ContextChanger initialLoggedIn={false}>
          <LoginStatusDisplay />
        </ContextChanger>,
      );

      expect(asFragment).toMatchSnapshot();

      const loginButton = screen.getByText('Login');
      expect(loginButton).toBeVisible();
      expect(loginButton.tagName).toBe('BUTTON');
    });

    it('should navigate to login route when the login button is pressed', async () => {
      render(
        <ContextChanger initialLoggedIn={false}>
          <LoginStatusDisplay />
        </ContextChanger>,
      );

      const loginButton = screen.getByText('Login');
      expect(loginButton).toBeInTheDocument();

      await act(async () => {
        loginButton.click();
      });

      expect(mockNavigateToRoute).toHaveBeenCalledWith('/login');
    });
  });

  describe('when logged in', () => {
    it('should show the logout button', async () => {
      const { asFragment } = render(
        <ContextChanger initialLoggedIn={true}>
          <LoginStatusDisplay />
        </ContextChanger>,
      );
      expect(asFragment).toMatchSnapshot();

      const logoutButton = screen.getByText('Logout');
      expect(logoutButton).toBeVisible();
      expect(logoutButton.tagName).toBe('BUTTON');
    });

    it('should logout when the logout button is pressed', async () => {
      const logoutPromise = PromiseWithResolvers();
      mockLogout.mockReturnValue(logoutPromise.promise);

      render(
        <ContextChanger initialLoggedIn={true}>
          <LoginStatusDisplay />
        </ContextChanger>,
      );

      let loginForm = screen.queryByText('Showing Log In Form');
      expect(loginForm).toBeNull();

      const logoutButton = screen.getByText('Logout');

      await act(async () => {
        logoutButton.click();
      });

      await act(async () => {
        logoutPromise.resolve({});
      });

      expect(mockLogout).toHaveBeenCalledTimes(1);

      loginForm = screen.queryByText('Showing Log In Form');
      expect(loginForm).toBeNull();
    });
  });
});
