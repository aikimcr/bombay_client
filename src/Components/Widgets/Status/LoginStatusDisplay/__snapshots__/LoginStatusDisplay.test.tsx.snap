// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LoginStatusDisplay when logged in should show the logout button 1`] = `[Function]`;

exports[`LoginStatusDisplay when logged out should show the login button 1`] = `[Function]`;
